### ISSUE TITLE
Floating point exception in module_sf_noahmp_glacier.F90

### ISSUE THREAD

**Issue by MicroTed** [ufs-community/ccpp-physics#114]
# Description

Potential division by zero (or 0/0) at line 2656 when dzsnso(0) = 0 [and snice(0)=0]. In debug compile of UFS model/FV3. Print statements show that isnow=0 at the (first) point where it fails and sneqv is just barely exceeding mwd. 

module_sf_noahmp_glacier.F90, lines 2653-2661
```
!to obtain equilibrium state of snow in glacier region
       
   if(sneqv > mwd) then   ! 100 mm -> maximum water depth
      bdsnow      = snice(0) / dzsnso(0)
      snoflow     = (sneqv - mwd)
      snice(0)    = snice(0)  - snoflow 
      dzsnso(0)   = dzsnso(0) - snoflow/bdsnow
      snoflow     = snoflow / dt
   end if
```

## Steps to Reproduce

Please provide detailed steps for reproducing the issue.

The RT rrfs_v1nssl test in debug compile fails with an FPE (tested on hera with intel compilers)
./opnReqTest -n rrfs_v1nssl -c dbg -ek

## Additional Context

Curiously, the rrfs_v1beta SDF does not trigger this failure, where the only difference is the microphysics option (Thompson instead of NSSL). I haven't found anything yet why that may be. When isnow=0, dzsnso is still initialized to zero but doesn't get updated in snowh2o_glacier. I don't know this code at all, so I hope somebody with knowledge can help isolate a possible cause (or just a patch like checking that dzsnso(0) /= 0 and snice(0) /= 0, or just isnow /= 0).

Presumably (I haven't checked), with an optimized compile, a NaN would be produced but doesn't cause any other harm.

## Output
On hera:

/scratch1/BMC/gsd-fv3-dev/Ted.Mansell/ufs/rt/stmp2/Ted.Mansell/FV3_OPNREQ_TEST/opnReqTest_24904/rrfs_v1nssl_intel_dbg_base/err:

```
 46: [h9c28:309589:0:309589] Caught signal 8 (Floating point exception: floating-point invalid operation)
 46: ==== backtrace (tid: 309589) ====
 46:  0 0x000000000004d455 ucs_debug_print_backtrace()  ???:0
 46:  1 0x000000000c33bf83 noahmp_glacier_routines_mp_snowwater_glacier_()  /scratch1/BMC/gsd-fv3-dev/Ted.Mansell/ufs/ufs-nssl3m/FV3/ccpp/physics/physics/module_sf_noahmp_glacier.F90:2656
 46:  2 0x000000000c339822 noahmp_glacier_routines_mp_water_glacier_()  /scratch1/BMC/gsd-fv3-dev/Ted.Mansell/ufs/ufs-nssl3m/FV3/ccpp/physics/physics/module_sf_noahmp_glacier.F90:2518
```

**Comment by MicroTed** [comment]
@barlage @cenlinhe I see you are listed as points of contact for Noah-MP. Maybe one of you has an idea on this? Thanks!

**Comment by dustinswales** [comment]
@MicroTed I was able to replicate this failure on Hera, /scratch1/BMC/gmtb/Dustin.Swales/stmp2/Dustin.Swales/FV3_OPNREQ_TEST/opnReqTest_2832/rrfs_v1nssl_intel_dbg_base/err.
But I also got the same error when I ran with the rrfs_v1beta SDF, /scratch1/BMC/gmtb/Dustin.Swales/stmp2/Dustin.Swales/FV3_OPNREQ_TEST/opnReqTest_13715/rrfs_v1beta_intel_dbg_base/err?
Perhaps this problem is more invasive than thought?

**Comment by MicroTed** [comment]
@dustinswales Oh, that's interesting. Thanks for checking that -- I don't think I ran the opnReqTest on hera for v1beta. I did run manual tests on jet with both compiled for debug. On jet only the v1nssl test had the FPE, where I used fewer cores than the default test (layout 3,6 instead of 3,8). But it's good to know that it doesn't seem to matter which microphysics is used because that is what I'd expect.
Maybe a simple fix would be something like

if(sneqv > mwd .and. isnow /= 0) then

But I don't know the code enough to know if that would have other consequences. It looks to me that the layer will always be zero if isnow=0?

**Comment by dustinswales** [comment]
@MicroTed Yeah, we could add some simple guards against this. But like you, I'm not an expert on coupling to the LSM, so I defer to @barlage @cenlinhe on the best way to handle this, since it doesn't appear to be a MP issue.

**Comment by MicroTed** [comment]
I'll just note that the current ncar/ccpp main branch of this code has a much larger (hard-coded) threshold on sneqv  (2000 instead of mwd =100). It may or may not make a difference.

**Comment by grantfirl** [comment]
Returning to this issue, we've run into a different trigger for this issue in https://github.com/ufs-community/ccpp-physics/pull/120. For this case, I tried @MicroTed's suggestion of using

`if(sneqv > mwd .and. isnow /= 0) then`

and it did prevent the runtime error for the particular test (rrfs_v1beta_debug_intel) where the runtime error occurred.
